File: src/store.js
/**
 * In-memory doublets store.
 *
 * The associative data model used by LinksQL is built from a single primitive:
 * the **link** (a doublet). Every link is a triple
 *
 *   (index: source target)
 *
 * where `index` is the link's own unique identity and `source`/`target` are the
 * indices of the linked links. A link whose source and target both equal its own
 * index is a *point* — the atomic node of the graph.
 *
 * The store owns identity allocation and enforces the associative invariant that
 * a `(source, target)` pair identifies at most one link (deduplication). It is a
 * pure data container: pattern matching and the substitution operation live in
 * `substitution.js`, which only relies on the read/write surface defined here.
 */

/** Error thrown when a store operation would violate model integrity. */
export class LinkIntegrityError extends Error {
  /** @param {string} message - Human readable description. */
  constructor(message) {
    super(message);
    this.name = 'LinkIntegrityError';
  }
}

/**
 * Build the deduplication key for a `(source, target)` pair.
 *
 * @param {number} source - Source index.
 * @param {number} target - Target index.
 * @returns {string} Map key.
 */
function pairKey(source, target) {
  return `${source},${target}`;
}

/** A mutable set of links addressed by integer identity. */
export class LinksStore {
  constructor() {
    /** @type {Map<number, {index: number, source: number, target: number}>} */
    this.links = new Map();
    /** @type {Map<string, number>} */
    this.byPair = new Map();
    /** Next identity to hand out; kept above every used index. */
    this.nextIndex = 1;
  }

  /** @returns {number} Number of stored links. */
  get size() {
    return this.links.size;
  }

  /**
   * @param {number} index - Identity to test.
   * @returns {boolean} Whether a link with this identity exists.
   */
  has(index) {
    return this.links.has(index);
  }

  /**
   * @param {number} index - Identity to fetch.
   * @returns {{index: number, source: number, target: number}|undefined} Link.
   */
  get(index) {
    return this.links.get(index);
  }

  /**
   * Look up a link by its `(source, target)` structure.
   *
   * @param {number} source - Source index.
   * @param {number} target - Target index.
   * @returns {{index: number, source: number, target: number}|undefined} Link.
   */
  findByPair(source, target) {
    const index = this.byPair.get(pairKey(source, target));
    return index === undefined ? undefined : this.links.get(index);
  }

  /** @returns {Array<{index: number, source: number, target: number}>} All links. */
  all() {
    return [...this.links.values()];
  }

  /**
   * Reserve a fresh identity, advancing the allocator past it.
   *
   * @returns {number} A previously-unused index.
   */
  allocateIndex() {
    while (this.links.has(this.nextIndex)) {
      this.nextIndex += 1;
    }
    return this.nextIndex++;
  }

  /**
   * Keep the allocator above an externally chosen identity.
   *
   * @param {number} index - An index that is now in use.
   */
  reserveIndex(index) {
    if (index >= this.nextIndex) {
      this.nextIndex = index + 1;
    }
  }

  /**
   * Create (or, by deduplication, return) a link.
   *
   * @param {object} spec - Link specification.
   * @param {number} [spec.index] - Explicit identity; auto-allocated when absent.
   * @param {number} spec.source - Source index.
   * @param {number} spec.target - Target index.
   * @returns {{index: number, source: number, target: number}} The link.
   */
  create({ index, source, target }) {
    if (!Number.isInteger(source) || !Number.isInteger(target)) {
      throw new LinkIntegrityError('Link source and target must be integers');
    }
    const existing = this.findByPair(source, target);
    if (existing) {
      if (index !== undefined && index !== existing.index) {
        throw new LinkIntegrityError(
          `Link (${source} ${target}) already exists as ${existing.index}, ` +
            `cannot also be ${index}`
        );
      }
      return existing;
    }
    let id = index;
    if (id === undefined) {
      id = this.allocateIndex();
    } else {
      if (!Number.isInteger(id) || id < 1) {
        throw new LinkIntegrityError('Link index must be a positive integer');
      }
      if (this.links.has(id)) {
        throw new LinkIntegrityError(`Link index ${id} is already in use`);
      }
      this.reserveIndex(id);
    }
    const link = { index: id, source, target };
    this.links.set(id, link);
    this.byPair.set(pairKey(source, target), id);
    return link;
  }

  /**
   * Replace the structure of an existing link, preserving its identity unless a
   * new identity is requested.
   *
   * @param {number} index - Identity of the link to update.
   * @param {object} spec - New structure.
   * @param {number} spec.source - New source index.
   * @param {number} spec.target - New target index.
   * @param {number} [spec.newIndex] - Optional new identity (re-index).
   * @returns {{index: number, source: number, target: number}} Updated link.
   */
  update(index, { source, target, newIndex }) {
    const current = this.links.get(index);
    if (!current) {
      throw new LinkIntegrityError(`Cannot update missing link ${index}`);
    }
    const id = newIndex === undefined ? index : newIndex;
    const collision = this.findByPair(source, target);
    if (collision && collision.index !== index) {
      throw new LinkIntegrityError(
        `Cannot update link ${index}: (${source} ${target}) already exists ` +
          `as ${collision.index}`
      );
    }
    if (id !== index && this.links.has(id)) {
      throw new LinkIntegrityError(`Link index ${id} is already in use`);
    }
    this.byPair.delete(pairKey(current.source, current.target));
    this.links.delete(index);
    const link = { index: id, source, target };
    this.links.set(id, link);
    this.byPair.set(pairKey(source, target), id);
    this.reserveIndex(id);
    return link;
  }

  /**
   * Remove a link by identity.
   *
   * @param {number} index - Identity to remove.
   * @returns {boolean} Whether a link was removed.
   */
  delete(index) {
    const current = this.links.get(index);
    if (!current) {
      return false;
    }
    this.byPair.delete(pairKey(current.source, current.target));
    this.links.delete(index);
    return true;
  }

  /** Remove every link and reset identity allocation. */
  clear() {
    this.links.clear();
    this.byPair.clear();
    this.nextIndex = 1;
  }
}

File: rust/src/store.rs
//! In-memory doublets store.
//!
//! The associative data model used by LinksQL is built from a single primitive:
//! the **link** (a doublet). Every link is a triple
//!
//! ```text
//! (index: source target)
//! ```
//!
//! where `index` is the link's own unique identity and `source`/`target` are the
//! indices of the linked links. A link whose source and target both equal its
//! own index is a *point* — the atomic node of the graph.
//!
//! The store owns identity allocation and enforces the associative invariant
//! that a `(source, target)` pair identifies at most one link (deduplication).
//! It is a pure data container: pattern matching and the substitution operation
//! live in [`crate::substitution`], which only relies on the read/write surface
//! defined here.

use crate::{Error, Result};
use std::collections::HashMap;

/// A link (doublet): a triple of integer identities.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Link {
    /// The link's own unique identity.
    pub index: i64,
    /// The index this link points from.
    pub source: i64,
    /// The index this link points to.
    pub target: i64,
}

/// A mutable set of links addressed by integer identity.
///
/// Insertion order is preserved by [`LinksStore::all`] (matching the JavaScript
/// `Map` iteration order) via a parallel order list; links are *not* sorted by
/// index.
#[derive(Debug, Default, Clone)]
pub struct LinksStore {
    /// Identity → link.
    links: HashMap<i64, Link>,
    /// Insertion order of identities (mirrors JS `Map` key order).
    order: Vec<i64>,
    /// `(source, target)` → identity, for structural deduplication.
    by_pair: HashMap<(i64, i64), i64>,
    /// Next identity to hand out; kept above every used index.
    next_index: i64,
}

impl LinksStore {
    /// Create an empty store with the allocator starting at `1`.
    #[must_use]
    pub fn new() -> Self {
        Self {
            links: HashMap::new(),
            order: Vec::new(),
            by_pair: HashMap::new(),
            next_index: 1,
        }
    }

    /// Number of stored links.
    #[must_use]
    pub fn size(&self) -> usize {
        self.links.len()
    }

    /// Whether a link with this identity exists.
    #[must_use]
    pub fn has(&self, index: i64) -> bool {
        self.links.contains_key(&index)
    }

    /// Fetch a link by identity.
    #[must_use]
    pub fn get(&self, index: i64) -> Option<Link> {
        self.links.get(&index).copied()
    }

    /// Look up a link by its `(source, target)` structure.
    #[must_use]
    pub fn find_by_pair(&self, source: i64, target: i64) -> Option<Link> {
        self.by_pair
            .get(&(source, target))
            .and_then(|index| self.links.get(index).copied())
    }

    /// All links, in insertion order.
    #[must_use]
    pub fn all(&self) -> Vec<Link> {
        self.order.iter().map(|index| self.links[index]).collect()
    }

    /// Reserve a fresh identity, advancing the allocator past it.
    pub fn allocate_index(&mut self) -> i64 {
        while self.links.contains_key(&self.next_index) {
            self.next_index += 1;
        }
        let id = self.next_index;
        self.next_index += 1;
        id
    }

    /// Keep the allocator above an externally chosen identity.
    pub const fn reserve_index(&mut self, index: i64) {
        if index >= self.next_index {
            self.next_index = index + 1;
        }
    }

    /// Insert a link, recording insertion order and the pair index.
    fn insert(&mut self, link: Link) {
        self.links.insert(link.index, link);
        self.order.push(link.index);
        self.by_pair.insert((link.source, link.target), link.index);
    }

    /// Remove a link from the maps and the order list.
    fn remove(&mut self, index: i64) -> Option<Link> {
        let link = self.links.remove(&index)?;
        self.by_pair.remove(&(link.source, link.target));
        if let Some(pos) = self.order.iter().position(|&i| i == index) {
            self.order.remove(pos);
        }
        Some(link)
    }

    /// Create (or, by deduplication, return) a link.
    ///
    /// If `(source, target)` already exists, the existing link is returned (and a
    /// conflicting explicit `index` is an error). Otherwise the explicit `index`
    /// is used (it MUST be `>= 1` and unused) or a fresh one is allocated.
    pub fn create(&mut self, index: Option<i64>, source: i64, target: i64) -> Result<Link> {
        if let Some(existing) = self.find_by_pair(source, target) {
            if let Some(idx) = index {
                if idx != existing.index {
                    return Err(Error::LinkIntegrity(format!(
                        "Link ({source} {target}) already exists as {}, cannot also be {idx}",
                        existing.index
                    )));
                }
            }
            return Ok(existing);
        }
        let id = match index {
            None => self.allocate_index(),
            Some(idx) => {
                if idx < 1 {
                    return Err(Error::LinkIntegrity(
                        "Link index must be a positive integer".to_string(),
                    ));
                }
                if self.links.contains_key(&idx) {
                    return Err(Error::LinkIntegrity(format!(
                        "Link index {idx} is already in use"
                    )));
                }
                self.reserve_index(idx);
                idx
            }
        };
        let link = Link {
            index: id,
            source,
            target,
        };
        self.insert(link);
        Ok(link)
    }

    /// Replace the structure of an existing link, preserving its identity unless
    /// a new identity is requested.
    ///
    /// Errors if the link is missing, if the new `(source, target)` collides with
    /// a *different* link, or if a requested `new_index` is already in use.
    pub fn update(
        &mut self,
        index: i64,
        source: i64,
        target: i64,
        new_index: Option<i64>,
    ) -> Result<Link> {
        let Some(current) = self.links.get(&index).copied() else {
            return Err(Error::LinkIntegrity(format!(
                "Cannot update missing link {index}"
            )));
        };
        let id = new_index.unwrap_or(index);
        if let Some(collision) = self.find_by_pair(source, target) {
            if collision.index != index {
                return Err(Error::LinkIntegrity(format!(
                    "Cannot update link {index}: ({source} {target}) already exists as {}",
                    collision.index
                )));
            }
        }
        if id != index && self.links.contains_key(&id) {
            return Err(Error::LinkIntegrity(format!(
                "Link index {id} is already in use"
            )));
        }
        // Mirror the JavaScript `Map` semantics exactly: delete the old key then
        // set the new one. In JS, `map.delete(k); map.set(k, v)` re-appends the
        // entry even when the key is unchanged, so `update` always moves the link
        // to the end of iteration order.
        self.by_pair.remove(&(current.source, current.target));
        if let Some(pos) = self.order.iter().position(|&i| i == index) {
            self.order.remove(pos);
        }
        self.links.remove(&index);
        let link = Link {
            index: id,
            source,
            target,
        };
        self.links.insert(id, link);
        self.order.push(id);
        self.by_pair.insert((source, target), id);
        self.reserve_index(id);
        Ok(link)
    }

    /// Remove a link by identity. Returns whether a link was removed.
    pub fn delete(&mut self, index: i64) -> bool {
        self.remove(index).is_some()
    }

    /// Remove every link and reset identity allocation.
    pub fn clear(&mut self) {
        self.links.clear();
        self.order.clear();
        self.by_pair.clear();
        self.next_index = 1;
    }
}
